### ISSUE TITLE
Manual device placement and distributed frameworks

### ISSUE THREAD

**Issue by jacobwjs** [lucidrains/imagen-pytorch#17]
Hi just digging through the code, nice work once again.

I notice lot's of manual device placement, which is fairly abstracted but I'm not sure how it plays with the likes of Pytorch Lightning or Hugging face's Accelerator.

Have there been (from anyone) any tests doing distributed training using one of the above frameworks (not DDP)? 

Looking for feedback before getting started if anyone has already headed down that path :)

**Comment by leeskyed** [comment]
There is another repo that uses Hugging face's Accelerator.  
https://github.com/cene555/Imagen-pytorch

**Comment by jacobwjs** [comment]
> There is another repo that uses Hugging face's Accelerator. https://github.com/cene555/Imagen-pytorch

Thanks for the link. Will give it a look over.

**Comment by jacobwjs** [comment]
Really nice work by cene555, but I don't like the way the code reads for what I'd like to contribute (not a knock on his work btw :)).

I spent some time last night integrating Accelerate into this repo, and a few issues have popped up mostly from me trying to stay true to what is already in place.

1. Lot's of manual device placement, which lucidrains has done a great job of managing, but not sure how this changes as this repo grows. Lot's to keep track of, which if removed can be automagically handled by Accelerate and whatever backend is used (e.g. DDP, DeepSpeed, etc.)
2. Currently the most basic approach of using Accelerate w/ DDP makes the current strategy of training challenging. Grads are accumulated on a per-Unet basis, which DDP doesn't like as it tries to sync grads across all devices on a backwards() call. This leaves all Unets that aren't the current one having no grads and bombs.

A potential solution is to refactor training such that all Unets accumulate grads in their own `no_sync()` context and sync on the final micro-batch.

@lucidrains do you have any input on this?

**Comment by jacobwjs** [comment]
Quick update. I ended up pulling the Unets out of the Imagen module just to get this up and running, and it seems to be training across 4 GPUs using DDP.

To note, this warning pops up.
`[W reducer.cpp:347] Warning: Grad strides do not match bucket view strides. This may indicate grad was not created according to the gradient layout contract, or that the param's strides changed since DDP was constructed.  This is not an error, but may impair performance.
grad.sizes() = [64, 128, 1, 1], strides() = [128, 1, 128, 128]
bucket_view.sizes() = [64, 128, 1, 1], strides() = [128, 1, 1, 1] (function operator())`

I'm guessing there are non-contiguous inputs somewhere?

**Comment by lucidrains** [comment]
@jacobwjs hey Jacob, i'll probably get huggingface's accelerate integrated by next week's end

i'm guessing you are scaling to DDP because you are seeing good initial results?

**Comment by jacobwjs** [comment]
> @jacobwjs hey Jacob, i'll probably get huggingface's accelerate integrated by next week's end
> 
> i'm guessing you are scaling to DDP because you are seeing good initial results?

Hi Phil, glad to hear you're planning on integrating Accelerate. I'm currently doing the same and happy to help. Please let me know how to sync up efforts.

Yep, seeing behavior and initial results that warrant scaling up on larger datasets. I'd like to answer questions a bit quicker, thus the need for Accelerate integration.

Just to add to the Accelerate point, I think you'll run into a few points rather quickly. I'll try to be succinct.

For T5,
- Device placement needs to abstracted if using accelerate. https://github.com/lucidrains/imagen-pytorch/blob/bfe761b52c93f53c1a961c0912bed3b33042382c/imagen_pytorch/imagen_pytorch.py#L1577-L1578
- Calls to encode text will need an update. https://github.com/lucidrains/imagen-pytorch/blob/bfe761b52c93f53c1a961c0912bed3b33042382c/imagen_pytorch/t5.py#L55 <br> (maybe just pass in device from accelerate?)

For Imagen/Unets,
Accelerate is great, but it's pretty thin. Each backend framework has it's own set of intricacies. For instance,


**DDP** `DistributedType.MULTI_GPU`
- accelerate.prepare() can accept all modules/optimizers/dataloaders in one call.
- need to handle calls to modules once wrapped in DDP.
- need to manually handle gradient accumulation 'no_sync()'. e.g.,

                `if accelerator.distributed_type == DistributedType.MULTI_GPU:
                    # If using DDP we prevent models from synchronizing until we 
                    # have accumulated gradients.
                    #
                    if (step + 1) % gradient_accumulation_steps != 0:
                        with unets[unet_idx].no_sync():
                            loss = imagen(images, texts, unet_number = unet_number, unets = unets)
                            loss /= gradient_accumulation_steps
                            accelerator.backward(loss)
                    else:
                        # Sync grads on final example.
                        #
                        loss = imagen(images, texts, unet_number = unet_number, unets = unets)
                        loss /= gradient_accumulation_steps
                        accelerator.backward(loss)`



**DeepSpeed** `DistributedType.DEEPSPEED` (currently experimental support btw)
- Dataloaders and modules must be passed to `accelerate.prepare()` together.


**Fully Sharded Data Parallel** `DistributedType.FSDP`
- models must be prepared (i.e. accelerate.prepare()) before optimizers

And finally saving needs to be handled accordingly
`accelerator.wait_for_everyone()`
`unwrapped_model = accelerator.unwrap_model(model)`

I haven't touched TPU support yet, but I'm sure comes with it's own set of caveats.

**Comment by jacobwjs** [comment]
What does the roadmap look like for you? DDP support seems like a good start and, then moving to the others (deepspeed, fsdp, etc). What are your thoughts?

**Comment by lucidrains** [comment]
no worries, i'll take care of it

will start with DDP

**Comment by jacobwjs** [comment]
> no worries, i'll take care of it
> 
> will start with DDP

Sure. Looking forward to it.

**Comment by jacobwjs** [comment]
#73 closing this.
